### PR TITLE
146 implement deconnexion protocol

### PR DESCRIPTION
**Pull request by LucasTesnier**
Implement part 1/2 of the deconnection process :
- Create a new protocol 13 relative to a disconnection demand.
- Implement this new protocol on client and server side.
- Update the Notion RFC

Holding the Control C on client side:
- When a sigint is caught, the client asks the server to cut the communication.
- Server gets the demands and removes everything relative to the client of his memory.

A client can now connect multiple time to the same port :)

You can test this part by launching the game and play a bit, but Functional and Unit Testing will be implemented in the part 2/2 of this PR. (Like Timer process and Sigint gestion in server side)

### CODE REVIEW

File: Client/ClientRoom.cpp [hunk 1]
```diff
@@ -46,6 +46,17 @@ using namespace communicator_lib;
 using namespace client_data;
 using namespace ecs;
 
+static ClientRoom::ClientState *clientRoomState(nullptr);
+
+/// @brief Usefull function called when a sigint received.
```
Comment (osvegn) [inline Client/ClientRoom.cpp:52]: > Useful

File: Server/Room.cpp [hunk 2]
```diff
@@ -84,29 +86,63 @@ void Room::initEcsGameData(void)
...
     _worldInstance->addSystem<DecreaseLifeTime>();
+    _worldInstance->addSystem<DisconnectableSystem>();
 }
 
 void Room::startConnexionProtocol(void) { _communicatorInstance.get()->startReceiverListening(); }
 
 void Room::startLobbyLoop(void)
 {
-    CommunicatorMessage connexionDemand;
+    CommunicatorMessage connexionOperation;
```
Comment (ZQUEMA) [inline Server/Room.cpp:96]: connectionOperation

File: Server/Room.hpp [hunk 3]
```diff
@@ -68,6 +68,16 @@ namespace server_data
...
         unsigned short _remainingPlaces;
 
+        /// @brief Trait a disconnection request. Identify the player and add to it a disconnection component
+        /// @param communicatorMessage actual message data
+        void _holdADisconnectionRequest(CommunicatorMessage disconectionDemand);
+
+        /// @brief Cross all the player entity and return the matched one
+        /// @param networkId networkId of the player to match
+        /// @return founded entity id
+        /// @throw a ECSError if no player was found.
```
Comment (osvegn) [inline Server/Room.hpp:78]: > an ECSError

File: Client/ClientRoom.cpp [hunk 4]
```diff
@@ -46,6 +47,17 @@ using namespace communicator_lib;
...
```
Comment (osvegn) [inline Client/ClientRoom.cpp:52]: > Useful

File: Client/ClientRoom.cpp [hunk 5]
```diff
@@ -8,6 +8,7 @@
 /// @file Client/ClientRoom.cpp
 
 #include "ClientRoom.hpp"
+#include <csignal>
 #include <functional>
 #include "Error/Error.hpp"
 #include "GraphicECS/SFML/Components/ActionQueueComponent.hpp"
```
Comment (osvegn) [inline Client/ClientRoom.cpp:14]: Not in a good order ?

File: libs/R-TypeLogic/Server/Systems/DisconnectableSystem.hpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+#include "World/World.hpp"
+
+namespace ecs
+{
+    /// @brief This is the Disconnectable System.
+    /// It will add the death component to all the associated entity.
+    struct DisconnectableSystem : public System {
+        /// @brief Run function for the Disconnectable System
+        /// @param world The world where the collisions will be checked
```
Comment (osvegn) [inline libs/R-TypeLogic/Server/Systems/DisconnectableSystem.hpp:19]: which collision ?

**Review by ZQUEMA — CHANGES_REQUESTED** [review]
One minor change is requiered

**Review by osvegn — CHANGES_REQUESTED** [review]
Some missing things to change

File: Server/Room.cpp (hunk 2 above)
Comment (LucasTesnier) [reply Server/Room.cpp:96]: :eyes:

File: Client/ClientRoom.cpp (hunk 1 above)
Comment (LucasTesnier) [reply Client/ClientRoom.cpp:52]: :eyes:

File: Server/Room.hpp (hunk 3 above)
Comment (LucasTesnier) [reply Server/Room.hpp:78]: :eyes:

File: Client/ClientRoom.cpp (hunk 4 above)
Comment (LucasTesnier) [reply Client/ClientRoom.cpp:52]: Quentin :eyes:

File: Client/ClientRoom.cpp (hunk 5 above)
Comment (LucasTesnier) [reply Client/ClientRoom.cpp:14]: :eyes:

File: libs/R-TypeLogic/Server/Systems/DisconnectableSystem.hpp (hunk 6 above)
Comment (LucasTesnier) [reply libs/R-TypeLogic/Server/Systems/DisconnectableSystem.hpp:19]: :eyes:

File: libs/R-TypeLogic/Server/Systems/DisconnectableSystem.hpp [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+#ifndef DISCONNECTABLESYSTEM_HPP_
+#define DISCONNECTABLESYSTEM_HPP_
+
+#include "World/World.hpp"
+
+namespace ecs
+{
+    /// @brief This is the Disconnectable System.
+    /// It will add the death component to all the associated entity.
```
[the remark this answers is not in the document]
Comment (LucasTesnier) [reply libs/R-TypeLogic/Server/Systems/DisconnectableSystem.hpp:16]: :eyes:

File: Client/ClientRoom.cpp (hunk 5 above)
Comment (LucasTesnier) [reply Client/ClientRoom.cpp:14]: CI action is not okay with that, so sob sob

**Review by ZQUEMA — APPROVED** [review]

File: Client/ClientRoom.cpp (hunk 5 above)
Comment (osvegn) [inline Client/ClientRoom.cpp:14]: Not in good order

**Review by osvegn — APPROVED** [review]
Look good

**Pull request merged**